In [1]:
import tensorflow as tf


In [2]:
# Load data
mnist = tf.keras.datasets.mnist

In [3]:

# Divide into training and test
(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train, x_test = x_train / 255.0, x_test / 255.0

11490434/11490434 ━━━━━━━━━━━━━━━━━━━━ 1s 0us/step


In [5]:
x_train.shape

(60000, 28, 28)

In [6]:
# Build a model
model = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(input_shape=(28, 28)),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dropout(0.2),
  tf.keras.layers.Dense(10)
])

/usr/local/lib/python3.10/dist-packages/keras/src/layers/reshaping/flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


In [7]:
# Example prediction in logits/ log-odds scores, one for each class.
predictions = model(x_train[:1]).numpy()
predictions

array([[ 0.4511059 , -0.41159713, -0.7274936 ,  0.02216505, -0.5237576 ,
        -0.47533128,  0.07436992,  0.07120942, -0.60093695, -0.04344407]],
      dtype=float32)

In [8]:
# Probability of the prediction being 0, 1, 2, ..., 9
tf.nn.softmax(predictions).numpy()

array([[0.18237421, 0.07696553, 0.05611826, 0.11876182, 0.06879955,
        0.07221325, 0.12512644, 0.1247316 , 0.06368938, 0.11122005]],
      dtype=float32)

In [9]:
# Loss function
loss_fn = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)

In [10]:
# Loss at a prediction point
loss_fn(y_train[:1], predictions).numpy()

2.6281319

In [11]:
# Compiling the model
model.compile(optimizer='adam',
              loss=loss_fn,
              metrics=['accuracy'])

In [12]:
# Fit
model.fit(x_train, y_train, epochs=5)

Epoch 1/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 9s 4ms/step - accuracy: 0.8584 - loss: 0.4847
Epoch 2/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step - accuracy: 0.9544 - loss: 0.1513
Epoch 3/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 10s 3ms/step - accuracy: 0.9684 - loss: 0.1075
Epoch 4/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 11s 4ms/step - accuracy: 0.9735 - loss: 0.0876
Epoch 5/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 9s 5ms/step - accuracy: 0.9764 - loss: 0.0735


In [13]:
# Evaluate results
model.evaluate(x_test,  y_test, verbose=2)

313/313 - 1s - 2ms/step - accuracy: 0.9798 - loss: 0.0693


[0.06927751749753952, 0.9797999858856201]

In [14]:
# One can find the probability from the trained model, one can attach the softmax to the model
probability_model = tf.keras.Sequential([
        model,
        tf.keras.layers.Softmax()
    ])

In [15]:
# Now, the results on a sample
probability_model(x_test[:10])

<tf.Tensor: shape=(10, 10), dtype=float32, numpy=
array([[7.10200510e-10, 5.07157303e-08, 1.53573490e-06, 1.25338134e-04,
        3.36831322e-12, 8.66270511e-07, 6.20164185e-14, 9.99868155e-01,
        1.46857786e-07, 3.93947175e-06],
       [2.12424460e-08, 3.04895773e-04, 9.99426484e-01, 2.55133375e-04,
        2.92314591e-16, 1.17460559e-05, 9.40262964e-07, 2.73558129e-12,
        7.88596140e-07, 2.70594704e-14],
       [1.14193170e-07, 9.97694850e-01, 4.27089224e-04, 1.31600855e-05,
        8.44448805e-05, 8.53856727e-06, 2.45949068e-05, 1.57422572e-03,
        1.72389162e-04, 5.06675292e-07],
       [9.99929130e-01, 5.81452553e-10, 2.66139505e-05, 2.02148343e-07,
        6.83298680e-08, 1.30710293e-06, 3.40924889e-05, 5.93286677e-06,
        9.94434046e-09, 2.61676632e-06],
       [1.77772586e-06, 2.07004902e-09, 6.17825481e-06, 5.39993792e-08,
        9.97348189e-01, 3.80555548e-07, 2.37853828e-06, 7.49586034e-05,
        8.80525022e-07, 2.56514945e-03],
       [1.55835733e-09, 9

In [16]:
# connect google drive
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [17]:
customer_csv_path = '/content/drive/MyDrive/USC Coursework/Artificial Intelligence - CSCE 580/CSCE580-Fall2024-Nayeem-SharedInfo/customer_data.csv'
sales_csv_path = '/content/drive/MyDrive/USC Coursework/Artificial Intelligence - CSCE 580/CSCE580-Fall2024-Nayeem-SharedInfo/sales_data.csv'

In [19]:
from re import S
import pandas as pd

customer_data = pd.read_csv(customer_csv_path)
sales_data = pd.read_csv(sales_csv_path)

customer_data

customer_id  gender   age payment_method
0         C241288  Female  28.0    Credit Card
1         C111565    Male  21.0     Debit Card
2         C266599    Male  20.0           Cash
3         C988172  Female  66.0    Credit Card
4         C189076  Female  53.0           Cash
...           ...     ...   ...            ...
99452     C441542  Female  45.0    Credit Card
99453     C569580    Male  27.0           Cash
99454     C103292    Male  63.0     Debit Card
99455     C800631    Male  56.0           Cash
99456     C273973  Female  36.0    Credit Card

[99457 rows x 4 columns]

In [46]:
sales_data

invoice_no customer_id         category  quantity    price invoice_date  \
0        I138884     C241288         Clothing         5  1500.40   05-08-2022   
1        I317333     C111565            Shoes         3  1800.51   12-12-2021   
2        I127801     C266599         Clothing         1   300.08   09-11-2021   
3        I173702     C988172            Shoes         5  3000.85   16-05-2021   
4        I337046     C189076            Books         4    60.60   24-10-2021   
...          ...         ...              ...       ...      ...          ...   
99452    I219422     C441542         Souvenir         5    58.65   21-09-2022   
99453    I325143     C569580  Food & Beverage         2    10.46   22-09-2021   
99454    I824010     C103292  Food & Beverage         2    10.46   28-03-2021   
99455    I702964     C800631       Technology         4  4200.00   16-03-2021   
99456    I232867     C273973         Souvenir         3    35.19   15-10-2022   

          shopping_mall  
0                Kanyon  
1        Forum Istanbul  
2             Metrocity  
3          Metropol AVM  
4                Kanyon  
...                 ...  
99452            Kanyon  
99453    Forum Istanbul  
99454         Metrocity  
99455      Istinye Park  
99456  Mall of Istanbul  

[99457 rows x 7 columns]

In [45]:
# show data statistics for customer data
print(f"Data statistics:\n{customer_data.describe()}")

print(f'\nFeatures:\n{customer_data.columns}')

missing_values_per_column = customer_data.isna().sum()
print(f"\nMissing values per column:\n{missing_values_per_column}")

column_data_types = customer_data.dtypes
print(f"\nColumn data types:\n{column_data_types}")

for col in customer_data.columns:
    print(f"Data Type of '{col}': {type(customer_data.at[0, col])}")

print(f"\nDataframe shape:\n{customer_data.shape}")

Data statistics:
                age
count  99338.000000
mean      43.425859
std       14.989400
min       18.000000
25%       30.000000
50%       43.000000
75%       56.000000
max       69.000000

Features:
Index(['customer_id', 'gender', 'age', 'payment_method'], dtype='object')

Missing values per column:
customer_id         0
gender              0
age               119
payment_method      0
dtype: int64

Column data types:
customer_id        object
gender             object
age               float64
payment_method     object
dtype: object
Data Type of 'customer_id': <class 'str'>
Data Type of 'gender': <class 'str'>
Data Type of 'age': <class 'numpy.float64'>
Data Type of 'payment_method': <class 'str'>

Dataframe shape:
(99457, 4)


In [47]:
# show data statistics for customer data
print(f"Data statistics:\n{sales_data.describe()}")

print(f'\nFeatures:\n{sales_data.columns}')

missing_values_per_column = sales_data.isna().sum()
print(f"\nMissing values per column:\n{missing_values_per_column}")

# this isn't helpful
column_data_types = sales_data.dtypes
print(f"\nColumn data types:\n{column_data_types}")

for col in sales_data.columns:
    print(f"Data Type of '{col}': {type(sales_data.at[0, col])}")

print(f"\nDataframe shape:\n{sales_data.shape}")

Data statistics:
           quantity         price
count  99457.000000  99457.000000
mean       3.003429    689.256321
std        1.413025    941.184567
min        1.000000      5.230000
25%        2.000000     45.450000
50%        3.000000    203.300000
75%        4.000000   1200.320000
max        5.000000   5250.000000

Features:
Index(['invoice_no', 'customer_id', 'category', 'quantity', 'price',
       'invoice_date', 'shopping_mall'],
      dtype='object')

Missing values per column:
invoice_no       0
customer_id      0
category         0
quantity         0
price            0
invoice_date     0
shopping_mall    0
dtype: int64

Column data types:
invoice_no        object
customer_id       object
category          object
quantity           int64
price            float64
invoice_date      object
shopping_mall     object
dtype: object
Data Type of 'invoice_no': <class 'str'>
Data Type of 'customer_id': <class 'str'>
Data Type of 'category': <class 'str'>
Data Type of 'quantity': <cla

**b) 1. 119 datapoints missing in age column that's .11% data missing**

In [53]:
# unique category of products
print(f"Unique categories:\n{sales_data['category'].unique()}")
print(f"\nNumber of items bought per category:\n{sales_data['category'].value_counts()}")



Unique categories:
['Clothing' 'Shoes' 'Books' 'Cosmetics' 'Food & Beverage' 'Toys'
 'Technology' 'Souvenir']

Number of items bought per category:
category
Clothing           34487
Cosmetics          15097
Food & Beverage    14776
Toys               10087
Shoes              10034
Souvenir            4999
Technology          4996
Books               4981
Name: count, dtype: int64


**b) 2. 34487 number of Clothing items is the hightest number**

In [54]:
mod_df = pd.merge(customer_data, sales_data, on='customer_id')
mod_df

customer_id  gender   age payment_method invoice_no         category  \
0         C241288  Female  28.0    Credit Card    I138884         Clothing   
1         C111565    Male  21.0     Debit Card    I317333            Shoes   
2         C266599    Male  20.0           Cash    I127801         Clothing   
3         C988172  Female  66.0    Credit Card    I173702            Shoes   
4         C189076  Female  53.0           Cash    I337046            Books   
...           ...     ...   ...            ...        ...              ...   
99452     C441542  Female  45.0    Credit Card    I219422         Souvenir   
99453     C569580    Male  27.0           Cash    I325143  Food & Beverage   
99454     C103292    Male  63.0     Debit Card    I824010  Food & Beverage   
99455     C800631    Male  56.0           Cash    I702964       Technology   
99456     C273973  Female  36.0    Credit Card    I232867         Souvenir   

       quantity    price invoice_date     shopping_mall  
0             5  1500.40   05-08-2022            Kanyon  
1             3  1800.51   12-12-2021    Forum Istanbul  
2             1   300.08   09-11-2021         Metrocity  
3             5  3000.85   16-05-2021      Metropol AVM  
4             4    60.60   24-10-2021            Kanyon  
...         ...      ...          ...               ...  
99452         5    58.65   21-09-2022            Kanyon  
99453         2    10.46   22-09-2021    Forum Istanbul  
99454         2    10.46   28-03-2021         Metrocity  
99455         4  4200.00   16-03-2021      Istinye Park  
99456         3    35.19   15-10-2022  Mall of Istanbul  

[99457 rows x 10 columns]

In [75]:
print(f"\nFrequncy of Female and Male presence in the dataframe:\n{display(mod_df['gender'].value_counts())}")

gender
Female    59482
Male      39975
Name: count, dtype: int64


Frequncy of Female and Male presence in the dataframe:
None


In [67]:
# quantity of items Females and Males purchase

print(f"\nNumber of items purchased by Females  :  {mod_df[mod_df['gender'] == 'Female']['quantity'].sum()}")
print(f"\nNumber of items purchased by Males    :  {mod_df[mod_df['gender'] == 'Male']['quantity'].sum()}")


Number of items purchased by Females  :  178659

Number of items purchased by Males    :  120053


In [58]:
print(f"\nFemale purcahse habits:\n{mod_df[mod_df['gender'] == 'Female']['category'].value_counts()}")
print(f"\nMale purchase habits:\n{mod_df[mod_df['gender'] == 'Male']['category'].value_counts()}")


Female purcahse habits:
category
Clothing           20652
Cosmetics           9070
Food & Beverage     8804
Toys                6085
Shoes               5967
Souvenir            3017
Technology          2981
Books               2906
Name: count, dtype: int64

Male purchase habits:
category
Clothing           13835
Cosmetics           6027
Food & Beverage     5972
Shoes               4067
Toys                4002
Books               2075
Technology          2015
Souvenir            1982
Name: count, dtype: int64


**b) 3. More number of Females buy more number of stuffs (Female, Male) = (items_purchased_female, items_purchased_male) : <br>(59482, 39975) = (178659, 120053)**

In [74]:
print(f"Frequency of prefered payment methods:{display(mod_df['payment_method'].value_counts())}")

payment_method
Cash           44447
Credit Card    34931
Debit Card     20079
Name: count, dtype: int64

Frequency of prefered payment methods:None



**b) 4.Cash is used the most**

In [77]:
mod_df['date'] = pd.to_datetime(mod_df['invoice_date'], format='%d-%m-%Y')
mod_df

customer_id  gender   age payment_method invoice_no         category  \
0         C241288  Female  28.0    Credit Card    I138884         Clothing   
1         C111565    Male  21.0     Debit Card    I317333            Shoes   
2         C266599    Male  20.0           Cash    I127801         Clothing   
3         C988172  Female  66.0    Credit Card    I173702            Shoes   
4         C189076  Female  53.0           Cash    I337046            Books   
...           ...     ...   ...            ...        ...              ...   
99452     C441542  Female  45.0    Credit Card    I219422         Souvenir   
99453     C569580    Male  27.0           Cash    I325143  Food & Beverage   
99454     C103292    Male  63.0     Debit Card    I824010  Food & Beverage   
99455     C800631    Male  56.0           Cash    I702964       Technology   
99456     C273973  Female  36.0    Credit Card    I232867         Souvenir   

       quantity    price invoice_date     shopping_mall       date  
0             5  1500.40   05-08-2022            Kanyon 2022-08-05  
1             3  1800.51   12-12-2021    Forum Istanbul 2021-12-12  
2             1   300.08   09-11-2021         Metrocity 2021-11-09  
3             5  3000.85   16-05-2021      Metropol AVM 2021-05-16  
4             4    60.60   24-10-2021            Kanyon 2021-10-24  
...         ...      ...          ...               ...        ...  
99452         5    58.65   21-09-2022            Kanyon 2022-09-21  
99453         2    10.46   22-09-2021    Forum Istanbul 2021-09-22  
99454         2    10.46   28-03-2021         Metrocity 2021-03-28  
99455         4  4200.00   16-03-2021      Istinye Park 2021-03-16  
99456         3    35.19   15-10-2022  Mall of Istanbul 2022-10-15  

[99457 rows x 11 columns]

In [79]:
mod_df['day'] = mod_df['date'].dt.day
mod_df['month'] = mod_df['date'].dt.month_name()
mod_df['year'] = mod_df['date'].dt.year
mod_df

customer_id  gender   age payment_method invoice_no         category  \
0         C241288  Female  28.0    Credit Card    I138884         Clothing   
1         C111565    Male  21.0     Debit Card    I317333            Shoes   
2         C266599    Male  20.0           Cash    I127801         Clothing   
3         C988172  Female  66.0    Credit Card    I173702            Shoes   
4         C189076  Female  53.0           Cash    I337046            Books   
...           ...     ...   ...            ...        ...              ...   
99452     C441542  Female  45.0    Credit Card    I219422         Souvenir   
99453     C569580    Male  27.0           Cash    I325143  Food & Beverage   
99454     C103292    Male  63.0     Debit Card    I824010  Food & Beverage   
99455     C800631    Male  56.0           Cash    I702964       Technology   
99456     C273973  Female  36.0    Credit Card    I232867         Souvenir   

       quantity    price invoice_date     shopping_mall       date  day  \
0             5  1500.40   05-08-2022            Kanyon 2022-08-05    5   
1             3  1800.51   12-12-2021    Forum Istanbul 2021-12-12   12   
2             1   300.08   09-11-2021         Metrocity 2021-11-09    9   
3             5  3000.85   16-05-2021      Metropol AVM 2021-05-16   16   
4             4    60.60   24-10-2021            Kanyon 2021-10-24   24   
...         ...      ...          ...               ...        ...  ...   
99452         5    58.65   21-09-2022            Kanyon 2022-09-21   21   
99453         2    10.46   22-09-2021    Forum Istanbul 2021-09-22   22   
99454         2    10.46   28-03-2021         Metrocity 2021-03-28   28   
99455         4  4200.00   16-03-2021      Istinye Park 2021-03-16   16   
99456         3    35.19   15-10-2022  Mall of Istanbul 2022-10-15   15   

           month  year  
0         August  2022  
1       December  2021  
2       November  2021  
3            May  2021  
4        October  2021  
...          ...   ...  
99452  September  2022  
99453  September  2021  
99454      March  2021  
99455      March  2021  
99456    October  2022  

[99457 rows x 14 columns]

In [81]:
print(f"Month-wise frequency of datapoints in dataframe:\n{display(mod_df['month'].value_counts())}")

month
January      11608
February     10482
March         8730
July          7877
October       7764
May           7697
December      7680
August        7635
June          7581
November      7563
April         7487
September     7353
Name: count, dtype: int64

Month-wise frequency of datapoints in dataframe:
None


In [121]:
# how many items were purchased in a month
display(mod_df.groupby('month')['quantity'].sum())
dff = mod_df.groupby('month')['quantity'].sum()
print(f"Number of items purchased in each month:\n{dff.max()}")

month
April        22558
August       22921
December     23152
February     31343
January      34763
July         23699
June         22674
March        26224
May          23217
November     22586
October      23477
September    22098
Name: quantity, dtype: int64

Number of items purchased in each month:
34763


**b) 5. Month that appears most in the dataframe: January<br>Most quantity of items purchased in a month: January**

In [165]:
print('max price: ',mod_df['price'].max())

# $500 dollar bins from 0 to mod_df['price'].max() range
bins = list(range(0, int(mod_df['price'].max()) + 499, 500))
labels = [f"({bins[i]}-{bins[i+1]}]" for i in range(len(bins)-1)]

print(bins)

# bin-wise summation of the price
price_binned = mod_df.groupby(pd.cut(mod_df['price'], bins=bins))['price'].sum()
mod_df['price_bin'] = pd.cut(mod_df['price'], bins=bins, labels=labels)
print(display(price_binned))
print('max price: ',price_binned.max())

max price:  5250.0
[0, 500, 1000, 1500, 2000, 2500, 3000, 3500, 4000, 4500, 5000, 5500]


<ipython-input-165-836f3e75ce2a>:10: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  price_binned = mod_df.groupby(pd.cut(mod_df['price'], bins=bins))['price'].sum()


price
(0, 500]         5621141.66
(500, 1000]     11555800.16
(1000, 1500]    11709936.98
(1500, 2000]    14031801.10
(2000, 2500]     6954174.96
(2500, 3000]           0.00
(3000, 3500]     9219061.05
(3500, 4000]           0.00
(4000, 4500]     4204200.00
(4500, 5000]           0.00
(5000, 5500]     5255250.00
Name: price, dtype: float64

None
max price:  14031801.100000001


In [164]:
labels = [f"({bins[i]}-{bins[i+1]}]" for i in range(len(bins)-1)]
labels

['(0-500]',
 '(500-1000]',
 '(1000-1500]',
 '(1500-2000]',
 '(2000-2500]',
 '(2500-3000]',
 '(3000-3500]',
 '(3500-4000]',
 '(4000-4500]',
 '(4500-5000]',
 '(5000-5500]']

**b) 6.a The bin with the highest purchase is \$1500-\$2000**


In [166]:
print('Unique Customers:', mod_df['customer_id'].unique().size)
mod_df.groupby(pd.cut(mod_df['price'], bins=bins))['customer_id'].count()


# count = 0
# for i in range(0, mod_df.shape[0]):
#   if mod_df.at[i, 'price'] <= 500:
#     count += 1
# print(count)

Unique Customers: 99457


<ipython-input-166-5e7638162856>:2: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  mod_df.groupby(pd.cut(mod_df['price'], bins=bins))['customer_id'].count()


price
(0, 500]        56781
(500, 1000]     15801
(1000, 1500]     9879
(1500, 2000]     8950
(2000, 2500]     3022
(2500, 3000]        0
(3000, 3500]     3022
(3500, 4000]        0
(4000, 4500]     1001
(4500, 5000]        0
(5000, 5500]     1001
Name: customer_id, dtype: int64

**b) 6.b The bin with the most purchases is ($0-$500]**

In [167]:
dfk = mod_df.groupby(pd.cut(mod_df['price'], bins=bins))['month'].value_counts()
dfk

<ipython-input-167-7a6482383f3b>:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  dfk = mod_df.groupby(pd.cut(mod_df['price'], bins=bins))['month'].value_counts()


price         month   
(0, 500]      January     6664
              February    5979
              March       4953
              October     4417
              August      4409
                          ... 
(5000, 5500]  June          75
              May           75
              March         74
              April         73
              November      67
Name: count, Length: 132, dtype: int64

In [168]:
mod_df

customer_id  gender   age payment_method invoice_no         category  \
0         C241288  Female  28.0    Credit Card    I138884         Clothing   
1         C111565    Male  21.0     Debit Card    I317333            Shoes   
2         C266599    Male  20.0           Cash    I127801         Clothing   
3         C988172  Female  66.0    Credit Card    I173702            Shoes   
4         C189076  Female  53.0           Cash    I337046            Books   
...           ...     ...   ...            ...        ...              ...   
99452     C441542  Female  45.0    Credit Card    I219422         Souvenir   
99453     C569580    Male  27.0           Cash    I325143  Food & Beverage   
99454     C103292    Male  63.0     Debit Card    I824010  Food & Beverage   
99455     C800631    Male  56.0           Cash    I702964       Technology   
99456     C273973  Female  36.0    Credit Card    I232867         Souvenir   

       quantity    price invoice_date     shopping_mall       date  day  \
0             5  1500.40   05-08-2022            Kanyon 2022-08-05    5   
1             3  1800.51   12-12-2021    Forum Istanbul 2021-12-12   12   
2             1   300.08   09-11-2021         Metrocity 2021-11-09    9   
3             5  3000.85   16-05-2021      Metropol AVM 2021-05-16   16   
4             4    60.60   24-10-2021            Kanyon 2021-10-24   24   
...         ...      ...          ...               ...        ...  ...   
99452         5    58.65   21-09-2022            Kanyon 2022-09-21   21   
99453         2    10.46   22-09-2021    Forum Istanbul 2021-09-22   22   
99454         2    10.46   28-03-2021         Metrocity 2021-03-28   28   
99455         4  4200.00   16-03-2021      Istinye Park 2021-03-16   16   
99456         3    35.19   15-10-2022  Mall of Istanbul 2022-10-15   15   

           month  year    price_bin  
0         August  2022  (1500-2000]  
1       December  2021  (1500-2000]  
2       November  2021      (0-500]  
3            May  2021  (3000-3500]  
4        October  2021      (0-500]  
...          ...   ...          ...  
99452  September  2022      (0-500]  
99453  September  2021      (0-500]  
99454      March  2021      (0-500]  
99455      March  2021  (4000-4500]  
99456    October  2022      (0-500]  

[99457 rows x 15 columns]

In [169]:
binned_frequency = mod_df.groupby('month')['price_bin'].value_counts().unstack(fill_value=0)
binned_frequency['total'] = binned_frequency.sum(axis=1)
most_frequent_month = binned_frequency['total'].idxmax()

print("Month-wise bin frequency:")
print(binned_frequency)


Month-wise bin frequency:
price_bin  (0-500]  (500-1000]  (1000-1500]  (1500-2000]  (2000-2500]  \
month                                                                   
April         4246        1222          752          689          236   
August        4409        1163          757          661          232   
December      4395        1220          765          695          213   
February      5979        1702         1040          905          315   
January       6664        1821         1166         1041          366   
July          4400        1286          782          743          266   
June          4391        1168          745          663          222   
March         4953        1382          892          809          275   
May           4383        1200          774          701          234   
November      4360        1225          788          638          209   
October       4417        1195          744          739          232   
September     4184       

In [141]:
month_count = {}
count = 0
for i in range(0, mod_df.shape[0]):
  if mod_df.at[i, 'price'] <= 500:
    if mod_df.at[i, 'month'] in month_count:
      month_count[mod_df.at[i, 'month']] += 1
    else:
      month_count[mod_df.at[i, 'month']] = 1
    count += 1
print(count)

for k, i in month_count.items():
  print(k, i)

56781
November 4360
October 4417
March 4953
December 4395
July 4400
January 6664
February 5979
May 4383
April 4246
September 4184
June 4391
August 4409


**b) 6.c Januray has the highest number of items purchased**